File: p4/Deepspace/Weapon.rb
#encoding:utf-8
require_relative "WeaponType"
require_relative "WeaponToUI"
module Deepspace
    class Weapon
        def initialize(_name,_type,_uses)
            @name = _name
            @type = _type
            @uses = _uses
        end 
        def self.newCopy(w)
            Weapon.new(w.name,w.type,w.uses)
        end     

        attr_reader:name,:type,:uses

        def power()
            @type.power()
        end 
        def useIt()
            if (@uses>0)
                @uses = @uses - 1
                return power()
            else 
                return 1.0
            end     
        end 

        def to_s()
            return "Name: #{@name}, Type: #{@type}, Power: #{power()}, Uses: #{@uses}"
        end 

        def getUIversion()
			Deepspace::WeaponToUI.new(self)
		end      
    end 
    #a = Weapon.new("arma",WeaponType::LASER,2)
    #puts a
end 

File: p5/deepspace/src/deepspace/Weapon.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author Sergio
 */
public class Weapon implements CombatElement {
    
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String name, WeaponType type, int uses){
        this.name = name;
        this.type = type;
        this.uses = uses;
        
    }
    
    Weapon(Weapon s){
        this.name = s.name;
        this.type = s.type;
        this.uses = s.uses;
    }
    
    public WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
    
    WeaponType getType(){
        return this.type;      
    }
    
    @Override
    public int getUses(){
        return this.uses;
    }
    
    float power(){
        return this.type.getPower();
    }
    
    @Override
    public float useIt(){
        float ret;
        if(this.uses>0){
            --this.uses;
            ret = power();
        }else
            ret = 1.0f;
        return ret;        
    }
    
    @Override
    public String toString(){
        String out;
        out = "Name: " + name + ", Weapon Type: " + type + ", Uses: " + uses;
        return out;
    }
}
